### PR TITLE
GroupBy implementation

### PR DESCRIPTION
**Pull request by pgovind**
More work towards dotnet/corefx#26845

### CODE REVIEW

File: tests/Microsoft.Data.DataFrame.Tests/DataFrameTests.cs [hunk 1]
```diff
@@ -23,40 +23,80 @@ public static DataFrame MakeDataFrameWithTwoColumns(int length)
             return dataFrame;
         }
 
-        public static DataFrame MakeDataFrameWithAllColumnTypes(int length)
+        public static DataFrame MakeDataFrameForGroupBy(int length) => MakeDataFrameWithAllColumnTypes(length, true);
```
Comment (pgovind) [inline tests/Microsoft.Data.DataFrame.Tests/DataFrameTests.cs]: Just adding this for now. As I add more functions to the GroupBy class, this may set up a DataFrame with edge case-y data

File: src/Microsoft.Data/PrimitiveColumn.cs [hunk 2]
```diff
@@ -196,6 +230,31 @@ internal PrimitiveColumn<decimal> CloneAsDecimalColumn()
...
+                for (long i = 0; i < Length; i++)
+                {
+                    bool containsKey = multimap.TryGetValue(this[i] ?? default, out ICollection<long> values);
+                    if (containsKey)
+                    {
+                        values.Add(i);
+                    }
+                    else
+                    {
+                        multimap.Add(this[i] ?? default, new List<long>() { i });
```
Comment (pgovind) [inline src/Microsoft.Data/PrimitiveColumn.cs]: This would mean `Null` values in primitive columns will be equivalent to `0`. I *think* this is ok

File: src/Microsoft.Data/BaseColumn.cs [hunk 3]
```diff
@@ -59,16 +59,25 @@ protected set
             get => GetValue(startIndex, length);
         }
 
+        /// <summary>
+        /// Called internally from Merge and GroupBy
```
Comment (pgovind) [inline src/Microsoft.Data/BaseColumn.cs]: Merge does not exist yet, but it will soon

File: src/Microsoft.Data/DataFrame.cs [hunk 4]
```diff
@@ -260,6 +261,77 @@ public DataFrame Join(DataFrame other, string leftSuffix = "_left", string right
             return ret;
         }
 
+        public GroupBy GroupBy(string columnName)
+        {
+            int columnIndex = _table.GetColumnIndex(columnName);
+            if (columnIndex == -1)
+                throw new ArgumentException($"{columnName} does not exist");
```
Comment (pgovind) [inline src/Microsoft.Data/DataFrame.cs]: use Strings here

File: tests/Microsoft.Data.DataFrame.Tests/DataFrameTests.cs (hunk 1 above)
Comment (eerhardt) [reply tests/Microsoft.Data.DataFrame.Tests/DataFrameTests.cs]: Let's not introduce dead code into our tests. I don't see this method called anywhere.

Also having a `bool forGroupByTests` doesn't feel very sustainable. When the next test that comes along and needs some special test data, adding a new bool to these methods isn't going to work well.
Instead, it may be better to have common code in small reusable methods that are called from higher-level methods.

---
In reply to: [301800494](https://github.com/dotnet/corefxlab/pull/2696#discussion_r301800494) [](ancestors = 301800494)

File: src/Microsoft.Data/GroupBy.cs [hunk 5]
```diff
@@ -0,0 +1,432 @@
...
+                    count++;
+                }
+            });
+
+            EnumerateColumnsWithRows(groupByColumnDelegate, columnDelegate);
+            ret.SetTableRowCount(firstColumn.Length);
+            return ret;
+        }
+
+        private delegate object columnFunc(IEnumerable<long> rows);
```
Comment (eerhardt) [inline src/Microsoft.Data/GroupBy.cs]: Is this used?

File: tests/Microsoft.Data.DataFrame.Tests/DataFrameTests.cs [hunk 6]
```diff
@@ -744,5 +784,44 @@ public void TestJoin()
             Assert.Equal(join["Int_right"][2], right["Int"][2]);
             VerifyJoin(join, left, right, JoinAlgorithm.Inner);
         }
+
+        [Fact]
+        public void TestGroupBy()
+        {
+            DataFrame df = MakeDataFrameWithNumericAndBoolColumns(10);
+            DataFrame count = df.GroupBy("Bool").Count();
```
Comment (eerhardt) [inline tests/Microsoft.Data.DataFrame.Tests/DataFrameTests.cs]: Should we test grouping by something other than `Bool`?

File: src/Microsoft.Data/Microsoft.Data.DataFrame.csproj [hunk 7]
```diff
@@ -99,6 +99,11 @@
       <AutoGen>True</AutoGen>
       <DependentUpon>BaseColumn.Computations.tt</DependentUpon>
     </Compile>
+    <Compile Update="ColumnArithmeticTemplate.cs">
```
Comment (eerhardt) [inline src/Microsoft.Data/Microsoft.Data.DataFrame.csproj]: I don't think this file exists.  Typically `.ttinclude` files don't generate their own `.cs` file. (At least in my experience.)
This change can be reverted.

File: src/Microsoft.Data/DataFrame.cs [hunk 8]
```diff
@@ -260,6 +261,77 @@ public DataFrame Join(DataFrame other, string leftSuffix = "_left", string right
...
+
+            BaseColumn column = _table.Column(columnIndex);
+
+            switch (column)
```
Comment (eerhardt) [inline src/Microsoft.Data/DataFrame.cs]: How about instead of a massive switch like this (and not really being able to support new, extension column types), why don't we have a public abstract method on `BaseColumn` that gets called here?

File: src/Microsoft.Data/BaseColumn.cs (hunk 3 above)
Comment (eerhardt) [reply src/Microsoft.Data/BaseColumn.cs]: That isn't a very good public description of what the method does. Can we give it a better summary?

---
In reply to: [301802932](https://github.com/dotnet/corefxlab/pull/2696#discussion_r301802932) [](ancestors = 301802932)

File: src/Microsoft.Data/BaseColumn.cs [hunk 9]
```diff
@@ -59,16 +59,25 @@ protected set
...
+        public virtual void Resize(long length) => throw new NotImplementedException();
+
         /// <summary>
         /// Clone column to produce a copy potentially changing the order by supplying mapIndices and an invert flag
         /// </summary>
         /// <param name="mapIndices"></param>
         /// <param name="invertMapIndices"></param>
         /// <returns></returns>
         public virtual BaseColumn Clone(BaseColumn mapIndices = null, bool invertMapIndices = false, long numberOfNullsToAppend = 0) => throw new NotImplementedException();
+        public virtual BaseColumn Clone(IEnumerable<long> mapIndices, long numberOfNullsToAppend = 0) => throw new NotImplementedException();
```
Comment (eerhardt) [inline src/Microsoft.Data/BaseColumn.cs]: Do we need 2 `virtual` Clone methods? Is one a more general version of the other? It would be nice if we could figure out a way to only have a single `virtual Clone` method that takes the appropriate parameters.

File: src/Microsoft.Data/BaseColumn.Computations.cs [hunk 10]
```diff
@@ -33,41 +33,81 @@ public virtual void CumulativeMax()
             throw new NotImplementedException();
         }
 
+        public virtual void CumulativeMax(IEnumerable<long> rows)
```
Comment (eerhardt) [inline src/Microsoft.Data/BaseColumn.Computations.cs]: Is there a better name for this parameter other than `rows`? These aren't really the rows themselves - these are the rowIndices, right?

Comment (eerhardt) [inline src/Microsoft.Data/BaseColumn.Computations.cs]: Since these are all public methods, we should have XML comments on them.

File: src/Microsoft.Data/DataFrame.cs [hunk 11]
```diff
@@ -260,6 +261,77 @@ public DataFrame Join(DataFrame other, string leftSuffix = "_left", string right
...
+                    Dictionary<object, ICollection<long>> dictionary = column.HashColumnValues<object>();
+                    return new GroupBy<object>(this, columnIndex, dictionary);
+            }
+        }
+
+        // In a GroupBy call, columns get resized. We need to set the RowCount to reflect the true Length of the DataFrame. Internal only. Should not be exposed
+        internal void SetTableRowCount(long rowCount)
+        {
+            // Even if current RowCount == rowCount, do the validation
+            int numberOfColumns = ColumnCount;
```
Comment (eerhardt) [inline src/Microsoft.Data/DataFrame.cs]: (nit) Why a new variable here? Why not just use `ColumnCount` in the loop?

File: src/Microsoft.Data/GroupBy.cs [hunk 12]
```diff
@@ -0,0 +1,432 @@
...
+            });
+
+            EnumerateColumnsWithRows(groupByColumnDelegate, columnDelegate);
+            ret.SetTableRowCount(firstColumn.Length);
+
+            return ret;
+        }
+
+
+        public override DataFrame First()
```
Comment (eerhardt) [inline src/Microsoft.Data/GroupBy.cs]: (nit) should just be single new lines between methods.

File: src/Microsoft.Data/GroupBy.cs [hunk 13]
```diff
@@ -0,0 +1,432 @@
...
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+
+namespace Microsoft.Data
+{
+    /// <summary>
+    /// A GroupBy class that is typically the product of a DataFrame.GroupBy call
```
Comment (eerhardt) [inline src/Microsoft.Data/GroupBy.cs]: (minor) instead of saying `the product of a DataFrame.GroupBy call`, I would go with `the result of a DataFrame.GroupBy call`.

Also, add a period to the end of the sentence.

File: src/Microsoft.Data/GroupBy.cs [hunk 14]
```diff
@@ -0,0 +1,432 @@
...
+    public class GroupBy<TKey> : GroupBy
+    {
+        private int _groupByColumnIndex;
+        private Dictionary<TKey, ICollection<long>> _keyValuePairs;
+        private DataFrame _dataFrame;
+        private PrimitiveColumn<long> _mapIndices;
+
+        public GroupBy(DataFrame dataFrame, int groupByColumnIndex, Dictionary<TKey, ICollection<long>> keyValuePairs)
+        {
+            _groupByColumnIndex = groupByColumnIndex;
```
Comment (eerhardt) [inline src/Microsoft.Data/GroupBy.cs]: Since this is a public constructor - we should be verifying the parameters.

File: src/Microsoft.Data/GroupBy.cs [hunk 15]
```diff
@@ -0,0 +1,432 @@
...
+            _keyValuePairs = keyValuePairs;
+            _dataFrame = dataFrame;
+        }
+
+        public override DataFrame All()
+        {
+            throw new NotImplementedException();
```
Comment (eerhardt) [inline src/Microsoft.Data/GroupBy.cs]: Should we not add this method until it is actually implemented?

File: tests/Microsoft.Data.DataFrame.Tests/DataFrameTests.cs [hunk 16]
```diff
@@ -744,5 +784,44 @@ public void TestJoin()
...
+            Assert.Equal(2, count.RowCount);
+            Assert.Equal((long)5, count["Int"][0]);
```
Comment (eerhardt) [inline tests/Microsoft.Data.DataFrame.Tests/DataFrameTests.cs]: Should verify the 2nd row.

File: tests/Microsoft.Data.DataFrame.Tests/DataFrameTests.cs [hunk 17]
```diff
@@ -744,5 +784,44 @@ public void TestJoin()
...
+
+            DataFrame first = df.GroupBy("Bool").First();
+            Assert.Equal(2, first.RowCount);
```
Comment (eerhardt) [inline tests/Microsoft.Data.DataFrame.Tests/DataFrameTests.cs]: Should verify the values.

File: tests/Microsoft.Data.DataFrame.Tests/DataFrameTests.cs [hunk 18]
```diff
@@ -744,5 +784,44 @@ public void TestJoin()
...
+
+            DataFrame head = df.GroupBy("Bool").Head(3);
+            Assert.Equal(6, head.RowCount);
```
Comment (eerhardt) [inline tests/Microsoft.Data.DataFrame.Tests/DataFrameTests.cs]: Should verify the values. Repeated below.

File: tests/Microsoft.Data.DataFrame.Tests/DataFrameTests.cs [hunk 19]
```diff
@@ -744,5 +784,44 @@ public void TestJoin()
...
+
+            DataFrame max = df.GroupBy("Bool").Max();
+            Assert.Equal(2, max.RowCount);
+            Assert.Equal(8, max["Int"][0]);
+            Assert.Equal(9, max["Int"][1]);
+
+            DataFrame min = df.GroupBy("Bool").Min();
+            Assert.Equal(2, min.RowCount);
+            Assert.Equal(0, min["Int"][0]);
+            Assert.Equal(0, min["Int"][1]); // Because of the null value
```
Comment (eerhardt) [inline tests/Microsoft.Data.DataFrame.Tests/DataFrameTests.cs]: I'm not sure I get this comment. Can you elaborate more?

Also, if the behavior is surprising to warrant a comment, is there a bug here that should be addressed to make the behavior understandable?

File: src/Microsoft.Data/GroupBy.cs [hunk 20]
```diff
@@ -0,0 +1,432 @@
...
+        public abstract DataFrame Max();
+        public abstract DataFrame Min();
+        public abstract DataFrame Product();
+        public abstract DataFrame Sum();
+    }
+
+    public class GroupBy<TKey> : GroupBy
+    {
+        private int _groupByColumnIndex;
+        private Dictionary<TKey, ICollection<long>> _keyValuePairs;
```
Comment (eerhardt) [inline src/Microsoft.Data/GroupBy.cs]: `_keyValuePairs` doesn't seem like the best name for this variable. Does `_keysToRowIndices` describe it better?

File: src/Microsoft.Data/GroupBy.cs [hunk 21]
```diff
@@ -0,0 +1,432 @@
...
+        }
+
+        public override DataFrame All()
+        {
+            throw new NotImplementedException();
+        }
+
+        internal delegate void ColumnDelegate(int columnIndex, long rowIndex, IEnumerable<long> rows, TKey key, bool firstGroup);
+        internal delegate void GroupByColumnDelegate(long rowNumber, TKey key);
+        internal void EnumerateColumnsWithRows(GroupByColumnDelegate groupByColumnDelegate, ColumnDelegate columnDelegate)
```
Comment (eerhardt) [inline src/Microsoft.Data/GroupBy.cs]: Does this and the above delegates need to be `internal`? I think only this class uses them.

**Comment by pgovind** [comment]
Filed https://github.com/dotnet/corefxlab/issues/2700 as a follow up. I don't want to add to this patch

File: tests/Microsoft.Data.DataFrame.Tests/DataFrameTests.cs (hunk 19 above)
Comment (pgovind) [reply tests/Microsoft.Data.DataFrame.Tests/DataFrameTests.cs]: I filed this https://github.com/dotnet/corefxlab/issues/2700 as a follow up.

Consider a column of ints like [0, 1, 0, 1, null]. At the moment, this forms the groups : [0] and [1] instead of [0], [1] and [null]. `null` is considered equivalent to `default`. This is because the Dictionary used internally for grouping uses the following for it's key:
`dictionary.Add(column[rowIndex] ?? default, ...`

Once I fix https://github.com/dotnet/corefxlab/issues/2700, it should becomes more intuitive.

File: src/Microsoft.Data/PrimitiveColumn.cs (hunk 2 above)
Comment (pgovind) [reply src/Microsoft.Data/PrimitiveColumn.cs]: Filed https://github.com/dotnet/corefxlab/issues/2700 to address this. `null` should be it's own group

File: tests/Microsoft.Data.DataFrame.Tests/DataFrameTests.cs [hunk 22]
```diff
@@ -744,5 +772,156 @@ public void TestJoin()
...
+                    BaseColumn productColumn = product[originalColumn.Name];
+                    Assert.Equal("0", productColumn[r].ToString());
+
+                    BaseColumn sumColumn = sum[originalColumn.Name];
+                    Assert.Equal("20", sumColumn[r].ToString());
+                }
+            }
+        }
+
+        [Fact]
```
Comment (pgovind) [inline tests/Microsoft.Data.DataFrame.Tests/DataFrameTests.cs]: The template types are the ones changing here. Is there a way to use Theory and InlineData for this?

File: src/Microsoft.Data/DataFrame.cs (hunk 8 above)
Comment (eerhardt) [reply src/Microsoft.Data/DataFrame.cs]: >  I was just trying to avoid boxing 

What would be boxed? `BaseColumn` and `GropuBy` are classes and thus reference types. Calling a an abstract method on it won't box anything.

>  the default case is doing that right?

No, the default case is 

```C#
default:
                    Dictionary<object, ICollection<long>> dictionary = column.HashColumnValues<object>();
                    return new GroupBy<object>(this, columnIndex, dictionary);
```

Which means the dictionary is always a `Dictionary<object, ...>`.  If you made a method on BaseColumn that looked like 

```C#
public abstract GroupBy GroupBy(int columnIndex);
```

Then each column can decide how to create the `GroupBy` object and its underlying Dictionary.

File: src/Microsoft.Data/BaseColumn.Computations.cs [hunk 23]
```diff
@@ -12,67 +12,165 @@ namespace Microsoft.Data
...
+        /// </summary>
         public virtual object Max()
         {
             throw new NotImplementedException();
         }
 
+        /// <summary>
+        /// Returns the maximum of the values at rowIndices
+        /// </summary>
+        public virtual object Max(IEnumerable<long> rows)
```
Comment (eerhardt) [inline src/Microsoft.Data/BaseColumn.Computations.cs]: The parameter here is still called `rows`, which is inconsistent with the summary comment and the methods above.

File: tests/Microsoft.Data.DataFrame.Tests/DataFrameTests.cs (hunk 22 above)
Comment (eerhardt) [reply tests/Microsoft.Data.DataFrame.Tests/DataFrameTests.cs]: One option would be to use `[Theory]` and `[MemberData]`. Something like:

```

public static IEnumerable<object[]> GroupByDataFrames()
{
    MakeDataFrame<byte, bool>(10, false), 
    MakeDataFrame<char, bool>(10, false),
    ....
}

[Theory]
[MemberData(nameof(GroupByDataFrames))]
public void TestGoupByDifferentColumnTypes(DataFrame df)
{
    DataFrame grouped = df.GroupBy("Column1").Count();
    Assert.Equal(2, grouped.RowCount);
}

```

I'm not sure if it is significantly better than what you have here, but that would be a way to do it with a Theory.

---
In reply to: [304140369](https://github.com/dotnet/corefxlab/pull/2696#discussion_r304140369) [](ancestors = 304140369)

File: src/Microsoft.Data/StringColumn.cs [hunk 24]
```diff
@@ -255,5 +256,30 @@ private StringColumn Clone(PrimitiveColumn<long> mapIndices = null, bool invertM
             }
             return ret;
         }
+
+        public override Dictionary<TKey, ICollection<long>> HashColumnValues<TKey>()
```
Comment (eerhardt) [inline src/Microsoft.Data/StringColumn.cs]: `HashColumnValues` => what does "hash" have to do with this method? My understanding of the method is that it builds a map of values => rowIndices.  It doesn't really "hash" the values.

File: src/Microsoft.Data/GroupBy.cs [hunk 25]
```diff
@@ -0,0 +1,421 @@
...
+                firstColumn.Resize(rowIndex + 1);
+                firstColumn[rowIndex] = key;
+            });
+            ColumnDelegate columnDelegate = new ColumnDelegate((int columnIndex, long rowIndex, IEnumerable<long> rowEnumerable, TKey key, bool firstGroup) =>
+            {
+                if (columnIndex == _groupByColumnIndex)
+                    return;
+                BaseColumn column = _dataFrame.Column(columnIndex);
+                long count = 0;
+                IEnumerator<long> rows = rowEnumerable.GetEnumerator();
```
Comment (eerhardt) [inline src/Microsoft.Data/GroupBy.cs]: Why not use `foreach` here?

File: src/Microsoft.Data/GroupBy.cs [hunk 26]
```diff
@@ -0,0 +1,421 @@
...
+        /// <summary>
+        /// Return the first value in each group
+        /// </summary>
+        /// <returns></returns>
+        public abstract DataFrame First();
+
+        /// <summary>
+        /// Returns the first numberOfRows rows of each group
+        /// </summary>
+        /// <param name="numberOfRowsInEachGroup"></param>
```
Comment (eerhardt) [inline src/Microsoft.Data/GroupBy.cs]: This param name doesn't match the actual parameter. You should be getting warnings for these kinds of mismatches.

File: src/Microsoft.Data/strings.resx [hunk 27]
```diff
@@ -117,16 +117,37 @@
   <resheader name="writer">
     <value>System.Resources.ResXResourceWriter, System.Windows.Forms, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089</value>
   </resheader>
+  <data name="CannotResizeDown" xml:space="preserve">
+    <value>Cannot resize down</value>
+  </data>
   <data name="ColumnIndexOutOfRange" xml:space="preserve">
     <value>Index cannot be greater than the Column's Length</value>
   </data>
+  <data name="EmptyFile" xml:space="preserve">
```
Comment (eerhardt) [inline src/Microsoft.Data/strings.resx]: This isn't being used. Can you ensure all the resource strings are being used?

File: src/Microsoft.Data/DataFrame.cs (hunk 8 above)
Comment (pgovind) [reply src/Microsoft.Data/DataFrame.cs]: > What would be boxed? BaseColumn and GropuBy are classes and thus reference types. Calling a an abstract method on it won't box anything.

Right right, but that's not what I meant. So, the way it is set up at the moment, let's say I have a `PrimitiveColumn<int> intColumn`. The `GroupBy` method will call `intColumn.HashColumnValues<int>` to produce a `Dictionary<int...`. That's ok, no boxing involved. If instead, I had only the `default` case of `column.HashColumnValues<object>`(where `column` is really `intColumn` and 'object' is really `int`), that would involve boxing when going in and out of the dictionary right? Anyway, I set it up this way because I didn't want the `BaseColumn` objects to have a reference to their parent `DataFrame` object. With your API suggestion, maybe I can still do that. Let me see

File: src/Microsoft.Data/StringColumn.cs (hunk 24 above)
Comment (pgovind) [reply src/Microsoft.Data/StringColumn.cs]: Oh, I guess I associate dictionary with hashing. I changed it to "GroupColumnValues" now. Open to other names too.

File: src/Microsoft.Data/DataFrame.cs [hunk 28]
```diff
@@ -31,6 +32,10 @@ public DataFrame()
         public DataFrame(IList<BaseColumn> columns)
         {
             _table = new DataFrameTable(columns);
+            foreach (BaseColumn column in columns)
+            {
+                column.DataFrame = this;
```
Comment (pgovind) [inline src/Microsoft.Data/DataFrame.cs]: Is this really safe? `this` is being constructed here, so I'm not sure.

File: src/Microsoft.Data/BaseColumn.cs [hunk 29]
```diff
@@ -69,6 +84,10 @@ protected set
 
         public virtual BaseColumn Sort(bool ascending = true) => throw new NotImplementedException();
 
+        public virtual Dictionary<TKey, ICollection<long>> GroupColumnValues<TKey>() => throw new NotImplementedException();
```
Comment (eerhardt) [inline src/Microsoft.Data/BaseColumn.cs]: You probably don't need this public method anymore, now that `GroupBy` is a public method on column.

File: src/Microsoft.Data/DataFrame.cs (hunk 28 above)
Comment (eerhardt) [reply src/Microsoft.Data/DataFrame.cs]: Maybe instead of having `BaseColumn` hold onto the `DataFrame` for the lifetime, maybe passing in the `DataFrame` to `GroupBy()` would work better?

That way you could have a `Column` object without a DataFrame.

---
In reply to: [305510252](https://github.com/dotnet/corefxlab/pull/2696#discussion_r305510252) [](ancestors = 305510252)

File: src/Microsoft.Data/BaseColumn.cs (hunk 29 above)
Comment (pgovind) [reply src/Microsoft.Data/BaseColumn.cs]: Right, I thought of that and decided to keep it around for the `Merge` PR. I'm ok deleting the API on BaseColumn for now and revisiting later

**Comment by pgovind** [comment]
Thanks for the review and stamp @eerhardt
